File: server/models/notification.py
# _*_ coding : UTF-8 _*_
# @Time : 2025/12/28
# @Author : sonder
# @File : notification.py
# @Comment : 通知模型

from enum import IntEnum
from tortoise import fields
from models.common import BaseModel


class NotificationType(IntEnum):
    """通知类型"""
    LOGIN = 0       # 登录通知
    ANNOUNCEMENT = 1  # 全局公告
    MESSAGE = 2     # 系统消息


class NotificationScope(IntEnum):
    """通知范围"""
    ALL = 0         # 全部用户
    DEPARTMENT = 1  # 指定部门
    USER = 2        # 指定用户


class NotificationStatus(IntEnum):
    """通知状态"""
    DRAFT = 0       # 草稿
    PUBLISHED = 1   # 已发布
    REVOKED = 2     # 已撤回


class SystemNotification(BaseModel):
    """系统通知表"""
    
    title = fields.CharField(max_length=200, description="通知标题")
    content = fields.TextField(description="通知内容")
    type = fields.SmallIntField(default=2, description="通知类型：0登录通知 1全局公告 2系统消息")
    scope = fields.SmallIntField(default=0, description="通知范围：0全部 1指定部门 2指定用户")
    scope_ids = fields.JSONField(null=True, description="范围ID列表（部门ID或用户ID）")
    status = fields.SmallIntField(default=0, description="状态：0草稿 1已发布 2已撤回")
    priority = fields.SmallIntField(default=0, description="优先级：0普通 1重要 2紧急")
    publish_time = fields.DatetimeField(null=True, description="发布时间")
    expire_time = fields.DatetimeField(null=True, description="过期时间")
    creator = fields.ForeignKeyField(
        "system.SystemUser",
        related_name="created_notifications",
        on_delete=fields.SET_NULL,
        null=True,
        description="创建者"
    )
    
    class Meta:
        table = "system_notification"
        table_description = "系统通知表"
        ordering = ["-created_at"]


class UserNotification(BaseModel):
    """用户通知关联表（记录用户已读状态）"""
    
    notification = fields.ForeignKeyField(
        "system.SystemNotification",
        related_name="user_notifications",
        on_delete=fields.CASCADE,
        description="通知"
    )
    user = fields.ForeignKeyField(
        "system.SystemUser",
        related_name="notifications",
        on_delete=fields.CASCADE,
        description="用户"
    )
    is_read = fields.BooleanField(default=False, description="是否已读")
    read_time = fields.DatetimeField(null=True, description="阅读时间")
    
    class Meta:
        table = "user_notification"
        table_description = "用户通知关联表"
        unique_together = [("notification", "user")]
        ordering = ["-created_at"]

File: web/src/api/system/notification.ts
/**
 * 通知管理 API
 */

import request from '@/utils/http'

/** 通知类型 */
export enum NotificationType {
  LOGIN = 0,        // 登录通知
  ANNOUNCEMENT = 1, // 全局公告
  MESSAGE = 2       // 系统消息
}

/** 通知范围 */
export enum NotificationScope {
  ALL = 0,          // 全部用户
  DEPARTMENT = 1,   // 指定部门
  USER = 2          // 指定用户
}

/** 通知状态 */
export enum NotificationStatus {
  DRAFT = 0,        // 草稿
  PUBLISHED = 1,    // 已发布
  REVOKED = 2       // 已撤回
}

/** 通知优先级 */
export enum NotificationPriority {
  NORMAL = 0,       // 普通
  IMPORTANT = 1,    // 重要
  URGENT = 2        // 紧急
}

/** 通知信息 */
export interface NotificationInfo {
  id: string
  title: string
  content: string
  type: NotificationType
  scope: NotificationScope
  scope_ids: string[]
  status: NotificationStatus
  priority: NotificationPriority
  publish_time?: string
  expire_time?: string
  created_at: string
  updated_at: string
  creator_id?: string
  creator_name?: string
  statistics?: {
    total: number
    read: number
    unread: number
  }
}

/** 用户通知信息 */
export interface UserNotificationInfo {
  id: string
  notification_id: string
  title: string
  content: string
  notification_type: NotificationType
  priority: NotificationPriority
  publish_time: string
  creator_name: string
  is_read: boolean
  read_time?: string
  created_at: string
}

/** 创建通知参数 */
export interface CreateNotificationParams {
  title: string
  content: string
  type?: NotificationType
  scope?: NotificationScope
  scope_ids?: string[]
  priority?: NotificationPriority
  expire_time?: string
}

/** 更新通知参数 */
export interface UpdateNotificationParams {
  title?: string
  content?: string
  type?: NotificationType
  scope?: NotificationScope
  scope_ids?: string[]
  priority?: NotificationPriority
  expire_time?: string
}

// ==================== 通知管理 API ====================

/** 获取通知列表（管理端） */
export const fetchNotificationList = (params: {
  page?: number
  pageSize?: number
  type?: NotificationType
  status?: NotificationStatus
  title?: string
}) => {
  return request.get<{
    result: NotificationInfo[]
    total: number
    page: number
    pageSize: number
  }>({
    url: '/api/notification/list',
    params
  })
}

/** 获取通知详情 */
export const fetchNotificationInfo = (id: string) => {
  return request.get<NotificationInfo>({
    url: `/api/notification/info/${id}`
  })
}

/** 创建通知 */
export const createNotification = (data: CreateNotificationParams) => {
  return request.post<{ id: string }>({
    url: '/api/notification/create',
    data
  })
}

/** 更新通知 */
export const updateNotification = (id: string, data: UpdateNotificationParams) => {
  return request.put<null>({
    url: `/api/notification/update/${id}`,
    data
  })
}

/** 发布通知 */
export const publishNotification = (id: string) => {
  return request.post<{
    total_users: number
    online_count: number
    offline_count: number
  }>({
    url: `/api/notification/publish/${id}`
  })
}

/** 撤回通知 */
export const revokeNotification = (id: string) => {
  return request.post<null>({
    url: `/api/notification/revoke/${id}`
  })
}

/** 删除通知 */
export const deleteNotification = (id: string) => {
  return request.delete<null>({
    url: `/api/notification/delete/${id}`
  })
}

// ==================== 用户通知 API ====================

/** 获取我的通知列表 */
export const fetchMyNotifications = (params: {
  page?: number
  pageSize?: number
  is_read?: boolean
  type?: NotificationType
}) => {
  return request.get<{
    result: UserNotificationInfo[]
    total: number
    page: number
    pageSize: number
  }>({
    url: '/api/notification/my/list',
    params
  })
}

/** 标记通知已读 */
export const markNotificationRead = (id: string) => {
  return request.post<null>({
    url: `/api/notification/my/read/${id}`
  })
}

/** 全部标记已读 */
export const markAllNotificationsRead = () => {
  return request.post<null>({
    url: '/api/notification/my/read-all'
  })
}

/** 获取未读数量 */
export const fetchUnreadCount = () => {
  return request.get<{ count: number }>({
    url: '/api/notification/my/unread-count'
  })
}

/** 获取待推送通知（HTTP轮询） */
export const fetchPendingNotifications = () => {
  return request.get<{ notifications: any[] }>({
    url: '/api/notification/my/pending'
  })
}
